### PR TITLE
Capsule Layer (With Dynamic Routing Between Capsules) #51764

### PR DESCRIPTION
**Pull request by Silver-Whiskey**
This is an implementation of capsule neural networks as described in "Dynamic Routing Between Capsules" paper (https://arxiv.org/abs/1710.09829).

I made a capsule layer, which was not originally implemented in Tensorflow, which made the developers and researchers job a bit tedious.

I have implemented fuctions given below:
1.) init : where you can do all input-independent initialization
2.) build: where you know the shapes of the input tensors and can do the rest of the initialization
3.) call: where you know the shapes of the input tensors and can do the rest of the initialization
4.) squash: an activation function discussed in the paper
5.) softmax: re-implemented softmax
6.) get_config: this function is important for storing the layer config while saving the model.
7.) compute_output_shape: This function helps in parsing the input dimension to the next layer

### CODE REVIEW

**Comment by innat** [comment]
@Silver-Whiskey I have some memorable working experiences with CapsNet and so I'm a bit excited to see PR on that. Thanks. However, I'm still not sure whether it's fit to be part of the keras core layers because of its downside (heavily slow training for inner loop logic).

One point to mention, I think you've inspired from this [caplayer](https://github.com/keras-team/keras-contrib/blob/master/keras_contrib/layers/capsule.py) implementation. I didn't inspect the complete code but I think we also need some other layers to fulfill, such as the [Mask](https://github.com/XifengGuo/CapsNet-Keras/blob/master/capsulelayers.py#L33) layer, [PrimaryCapsule](https://github.com/XifengGuo/CapsNet-Keras/blob/master/capsulelayers.py#L178) layer, etc. Please correct me if I'm missing something.

image

Some References:
- [Official](https://github.com/Sarasra/models)
- [CapsNet-Tensorflow](https://github.com/naturomics/CapsNet-Tensorflow)
- [CapsNet-Keras](https://github.com/XifengGuo/CapsNet-Keras)
- [CapsNet-PyTorch](https://github.com/gram-ai/capsule-networks)
- [Step-by-Stepy-CapsNet-PyTorch](https://nn.labml.ai/capsule_networks/index.html)

**Comment by fchollet** [comment]
Thanks for the PR! Capsule networks are an exciting research area. However, in this case, the layer doesn't meet [our criteria for inclusion](https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md#carefully-weigh-whether-a-new-feature-should-be-included) in the core Keras API: "we only add new objects that are already commonly used in the machine learning community".
